src/actions/ratesActions.js

```diff
@@ -35,54 +38,72 @@ import { reportErrorLog } from 'utils/common';
-    const { rates: { data: currentRates = {} } } = getState();
-    // $FlowFixMe: flow update to 0.122
-    const rates = { ...currentRates, ...newRates };
-    dispatch(saveDbAction('rates', { rates }, true));
-    dispatch({ type: UPDATE_RATES, payload: rates });
+export const setIsFetchingRatesAction = (isFetching: boolean) => ({
+  type: SET_FETCHING_RATES,
+  payload: isFetching,
+});
+
+export const setRatesAction = (
+  chain: Chain,
+  rates: RatesBySymbol,
+) => {
+  return async (dispatch: Dispatch, getState: GetState) => {
+    if (isEmpty(rates)) return;
+
+    dispatch({ type: SET_CHAIN_RATES, payload: { chain, rates } });
+
+    const updatedRatesPerChain = ratesPerChainSelector(getState());
+    await dispatch(saveDbAction('rates', { rates: updatedRatesPerChain }, true));
   };
 };
 
 export const fetchAssetsRatesAction = () => {
   return async (dispatch: Dispatch, getState: GetState) => {
-    const assetsBalances = assetsBalancesSelector(getState());
-    const supportedAssets = supportedAssetsPerChainSelector(getState());
-
-    const allAccountsCrossChainAssets = Object.keys(assetsBalances).reduce((combinedAssets, accountId) => {
-      const accountAssetsBalances = assetsBalances[accountId] ?? {};
+    const {
+      rates: { isFetching },
+      session: { data: { isOnline } },
+    } = getState();
 
-      const accountAssets = Object.keys(accountAssetsBalances).reduce((assets, chain) => {
-        const chainSupportedAssets = supportedAssets[chain] ?? [];
-        const walletAssets = accountAssetsBalances[chain]?.wallet ?? {};
-        const mapped = mapWalletAssetsBalancesIntoAssetsBySymbol(walletAssets, chainSupportedAssets);
-        // $FlowFixMe
-        return { ...assets, ...mapped };
-      }, {});
+    if (isFetching || !isOnline) return;
 
-      return { ...combinedAssets, ...accountAssets };
-    }, {});
+    dispatch(setIsFetchingRatesAction(true));
 
-    if (isEmpty(allAccountsCrossChainAssets)) {
-      reportErrorLog('fetchAssetsRatesAction failed: allAccountsCrossChainAssets is empty');
-      return;
-    }
+    const assetsBalances = assetsBalancesSelector(getState());
+    const supportedAssets = supportedAssetsPerChainSelector(getState());
 
-    const rates = await getExchangeRates(allAccountsCrossChainAssets);
-    dispatch(setRatesAction(rates));
+    await Promise.all(Object.keys(assetsBalances).map(async (accountId) => {
+      const accountAssetsBalances = assetsBalances[accountId] ?? {};
+      await Promise.all(Object.keys(accountAssetsBalances).map(async (chain) => {
+        try {
+          const chainSupportedAssets = supportedAssets[chain] ?? [];
+          const walletAssets = accountAssetsBalances[chain]?.wallet ?? {};
+
+          const accountAssets = mapWalletAssetsBalancesIntoAssetsBySymbol(walletAssets, chainSupportedAssets);
+          const rates = await getExchangeRates(chain, accountAssets);
+
+          await dispatch(setRatesAction(chain, rates));
```

Comment:
❗ Seems like we're replacing previous chain rates for user assets. This will cause problems if user has 2 accounts: 1st with assets A, B, C, 2nd with assets A, D, E. The last of promises to finish will set its chain rates resulting either in A, B, C rates or A, D, E while redux should contain all A, B, C, D, E.

Consider improving the logic:
```
for each chain:
  get all assets for given chain across accounts
  query exchange rates for that chain & all accounts assets
  set redux state

This should result in fewer network calls. and all needed rates in redux store.

```

Reply:
This has changed by replacing rates from setting to updating flow.

Reply:
Combined requested assets tho.